autogpt/memory/vectara.py

```diff
@@ -0,0 +1,291 @@
+
+    def get_jwt_token(self):
+        """Returns JWT that can be attached for authorization."""
+        from authlib.integrations.requests_client import OAuth2Session
+
+        # Refresh token every 5 mins.
+        if (time.time() - self._token_refresh_ts) < 5 * 60:
+            return self._token
+
+        logging.debug(
+            "Refreshing JWT token after %s secs.", time.time() - self._token_refresh_ts
+        )
+        session = OAuth2Session(self._app_id, self._app_secret, scope="")
+        token = session.fetch_token(
+            self._token_endpoint, grant_type="client_credentials"
+        )  # authorization_code
+
+        self._token_refresh_ts = time.time()
+        self._token = token["access_token"]
+        return self._token
+
+
+class VectaraMemory(MemoryProviderSingleton):
+    """A memory provider backed by Vectara (https://vectara.com)."""
+
+    def __init__(self, cfg):
+        self._jwt_provider = JWTTokenProvider(
+            cfg.vectara_customer_id, cfg.vectara_app_id, cfg.vectara_app_secret
+        )
+        self._customer_id = cfg.vectara_customer_id
+        self._corpus_id = self.get_corpus_id(cfg.memory_index)
+        logging.debug("Using corpus id %s", self._corpus_id)
+
+    def _get_post_headers(self):
+        """Returns headers that should be attached to each post request."""
+        return {
+            "Authorization": f"Bearer {self._jwt_provider.get_jwt_token()}",
+            "customer-id": str(self._customer_id),
+            "Content-Type": "application/json",
+        }
+
+    def get_corpus_id(self, index_name):
+        """Returns the corpus id for the passed 'index_name'.
+
+        Vectara APIs use int corpus (aka index) idss instead of string names. This
+        method can be used to get the int corpus id.
+        If a corpus with the name already exists, its id is returned. Otherwise, a new
+        corpus is created with the name, and its id is returned.
+        """
+        response = requests.post(
+            url="https://api.vectara.io/v1/list-corpora",
+            headers=self._get_post_headers(),
+            data=json.dumps({"numResults": 2, "filter": f"^{index_name}$"}),
+            timeout=15,
+        )
+        if response.status_code != 200:
+            logger.typewriter_log(
+                "FAILED TO LIST VECTARA CORPORA",
+                Fore.RED,
+                Style.BRIGHT
+                + (
+                    f"code {response.status_code}, reason {response.reason}, details {response.text}"
+                )
+                + Style.RESET_ALL,
+            )
+            logger.double_check(
+                "Please ensure you have setup and configured Vectara properly for use."
+                f"You can check out {Fore.CYAN + Style.BRIGHT}"
+                "https://github.com/Torantulino/Auto-GPT#-vectara-setup"
+                f"{Style.RESET_ALL} to ensure you've set up everything correctly."
+            )
```

Comment:
would love a bit more specific errors here. 403, 500, etc for common use cases so we don't get too many issues

Reply:
I've added more details to message error message here and elsewhere.